Update Error Message for Nothing Inputs to Graphs

clairelin135 opened the pull request:
Motivated by https://github.com/dagster-io/dagster/issues/7610.

I've also seen similar questions in dagster-support where users anticipate `Nothing` inputs to graphs don't need to be specified as input args (which is true for ops). I think it's worth clarifying this behavior and making the error message much more specific.

sryza commented:
Adding @alangenfeld because I am less familiar with this part of the world. Something to consider here is that supplying Dagster Types for graph inputs is only possible with the legacy APIs. Eventually, we'll deprecate the `input_defs` argument on `GraphDefinition`. The `ins` argument accepts `GraphIn`s, which do not have Dagster Types.

alangenfeld commented:
> supplying Dagster Types for graph inputs is only possible with the legacy APIs. Eventually, we'll deprecate the input_defs argument on GraphDefinition. The ins argument accepts GraphIns, which do not have Dagster Type

This is not how things currently work, type hints on the @graph decorated function will populate DagsterTypes in the created `InputDefinition` s . We should fix this if this is not how it is intended to work, though type loader based config setting will break for users if we change this behavior. Touched on here https://github.com/dagster-io/dagster/issues/7109

alangenfeld commented on python_modules/dagster/dagster/core/definitions/decorators/solid_decorator.py:
```diff
@@ -373,6 +374,20 @@ def resolve_checked_solid_fn_inputs(
 
     undeclared_inputs = explicit_names - used_inputs
+
+    if decorator_name == "@graph":
```
i would leave a little comment here explaining the context for why we raise a special error here before the other one below

alangenfeld commented on python_modules/dagster/dagster/core/definitions/decorators/solid_decorator.py:
```diff
@@ -373,6 +374,20 @@ def resolve_checked_solid_fn_inputs(
+        if undeclared_nothing_inputs:
+            undeclared_nothing_inputs_printed = ", ".join(undeclared_nothing_inputs)
+            raise DagsterInvalidDefinitionError(
+                f"Nothing-type input definitions declared in {decorator_name} "
+                "must be specified as input arguments to the decorated function."
```
nit: maybe a chance to add some explainer here that they will need to map the nothing to any ops they wish to have the dependency. Wording concisely may be tough but i think this is pretty much the best place to explain that to the user.

alangenfeld commented on python_modules/dagster/dagster/core/definitions/decorators/solid_decorator.py:
```diff
@@ -373,6 +374,20 @@ def resolve_checked_solid_fn_inputs(
+        undeclared_nothing_inputs = [
+            inp.name
+            for inp in explicit_input_defs
+            if inp.dagster_type.kind == DagsterTypeKind.NOTHING and inp.name in undeclared_inputs
```
can do dagster_type.is_nothing

alangenfeld replied:
though to sandys point - the actual type being set to Nothing is increasingly rare and if we change the APIs may never proc. 

maybe a better approach would be to have to split the error message we throw below for undeclared inputs into different messages for composition functions (graph/job) and compute function (op, resource, ). 

Taking a closer look at the message its kind of incorrect for graph/job anyway as it mentions the context parameter.

alangenfeld reviewed (approved):
nice, i think this is worth the small code cost - its a real good chance to help a user when they have encountered this un-intuitive part of the system

alangenfeld reviewed (changes requested):
ok on second thought lets try to make this an even better improvement by making all undeclared_inputs exceptions more clear

alangenfeld reviewed (commented):
>  I think we should probably stop doing this?

based on support inquiries I think the type-loader based config value setting gets used enough that we have to be careful not to regress that too bad if we remove it. I think thats the only concern though.

sryza commented:
> based on support inquiries I think the type-loader based config value setting gets used enough that we have to be careful not to regress that too bad if we remove it. I think thats the only concern though.

I agree that we have to be careful not to regress too bad. One way to do this would be to check if all the leaf ops inputs that are mapped to a graph input have the same dagster type and then use the loader from that dagster type. Though does raise the risk of possibly confusing error state if you forget to put it on one of those inputs.

alangenfeld commented:
ya i was thinking the same thing - tricky as you called out which almost makes me want to have a way to explicitly set it which gets us back to not removing it. worth more discussion - I can snapshot some of this on to the issue I linked above.

clairelin135 commented:
Thanks @alangenfeld and @sryza for the feedback. I've adjusted the error messaging to clarify the behavior in composition functions that nothing-type inputs must be specified as input args to the decorated function and passed to dependent ops. I've also added comments to explain this and fixed the error that mentions context for graphs.
